src/unix.rs

```diff
@@ -7,27 +7,204 @@
+use std::collections::HashSet;
+use std::io;
 
-use futures::future;
-use futures::stream::Fuse;
-use futures::sync::mpsc;
-use futures::sync::oneshot;
-use futures::{Future, Stream, IntoFuture, Poll, Async};
 use self::libc::c_int;
-use self::tokio_uds::UnixStream;
+use self::nix::sys::signal::{sigaction, SigAction, SigHandler, SigSet, SA_NOCLDSTOP, SA_RESTART};
+use self::nix::sys::signal::Signal as NixSignal;
+use self::nix::Error as NixError;
+use self::nix::Errno;
+use self::nix::sys::socket::{recv, send, socketpair, AddressFamily, SockType, SockFlag, MSG_DONTWAIT};
+use self::mio::{Evented, Token, Ready, PollOpt};
+use self::mio::Poll as MioPoll;
+use self::mio::unix::EventedFd;
+use futures::{Async, AsyncSink, Future, IntoFuture};
+use futures::sync::mpsc::{Receiver, Sender, channel};
+use futures::{Sink, Stream, Poll};
+use tokio_core::reactor::{Handle, CoreId, PollEvented};
 use tokio_core::io::IoFuture;
-use tokio_core::reactor::{PollEvented, Handle};
 
-static INIT: Once = ONCE_INIT;
-static mut GLOBAL_STATE: *mut GlobalState = 0 as *mut _;
+pub use self::libc::{SIGINT, SIGTERM, SIGUSR1, SIGUSR2};
+pub use self::libc::{SIGHUP, SIGQUIT, SIGPIPE, SIGALRM, SIGTRAP};
+
+// Number of different unix signals
+const SIGNUM: usize = 32;
+
+#[derive(Default)]
+struct SignalInfo {
+    initialized: bool,
+    // The ones interested in this signal
+    recipients: Vec<Sender<c_int>>,
+    // TODO: Other stuff, like the previous sigaction to call
+}
+
+struct Globals {
+    pending: [AtomicBool; SIGNUM],
+    sender: RawFd,
+    receiver: RawFd,
+    signals: [Mutex<SignalInfo>; SIGNUM],
+    drivers: Mutex<HashSet<CoreId>>,
+}
+
+impl Globals {
+    fn new() -> Self {
+        // TODO: Better error handling
+        // We use socket pair instead of pipe, as it allows send() and recv().
+        let (receiver, sender) = socketpair(AddressFamily::Unix, SockType::Stream, 0, SockFlag::empty()).unwrap();
+        Globals {
+            // Bunch of false values
+            pending: Default::default(),
+            sender: sender,
+            receiver: receiver,
+            signals: Default::default(),
+            drivers: Mutex::new(HashSet::new()),
+        }
+    }
+}
+
+lazy_static! {
+    // TODO: Get rid of lazy_static once the prototype is done ‒ get rid of the dependency as well
+    // as the possible lock in there, which *might* be problematic in signals
+    static ref GLOBALS: Globals = Globals::new();
+}
+
+// Flag the relevant signal and wake up through a self-pipe
+extern "C" fn pipe_wakeup(signal: c_int) {
+    let index = signal as usize;
+    // TODO: Handle the old signal handler
+    // It might be good enough to use some lesser ordering than this, but how to prove it?
+    GLOBALS.pending[index].store(true, Ordering::SeqCst);
+    // Send a wakeup, ignore any errors (anything reasonably possible is full pipe and then it will
+    // wake up anyway).
+    let _ = send(GLOBALS.sender, &[0u8], MSG_DONTWAIT);
```

Comment:
Tokio-uds is problematic, because it needs a handle it is created on. I guess nobody really promises anything about what happens to the socket once the loop is destroyed. And I don't know how to reasonably stick the receiver end into multiple loops.

I might probably use mio-uds, but it still doesn't give explicit guarantees about not doing some forbidden stuff in a signal handler (like memory allocation which might acquire a lock). But that one would probably be fine.